crates/polars-io/src/csv/parser.rs

```diff
@@ -1,11 +1,77 @@
+use std::path::PathBuf;
+
 use memchr::memchr2_iter;
 use num_traits::Pow;
 use polars_core::prelude::*;
+use polars_core::POOL;
+use polars_utils::index::Bounded;
+use rayon::prelude::*;
 
 use super::buffer::*;
 use crate::csv::read::NullValuesCompiled;
 use crate::csv::splitfields::SplitFields;
+use crate::csv::utils::get_file_chunks;
 use crate::csv::CommentPrefix;
+use crate::utils::get_reader_bytes;
+
+/// Read the number of rows without parsing columns
+/// useful for count(*) queries
+pub fn count_rows(
+    path: &PathBuf,
+    separator: u8,
+    quote_char: Option<u8>,
+    comment_prefix: Option<&CommentPrefix>,
+    eol_char: u8,
+    has_header: bool,
+) -> PolarsResult<usize> {
+    let mut reader = polars_utils::open_file(path)?;
+    let reader_bytes = get_reader_bytes(&mut reader)?;
+    const MIN_ROWS_PER_THREAD: usize = 1024;
+    let max_threads = POOL.current_num_threads();
+
+    // Determine if parallelism is beneficial and how many threads
+    let n_threads = get_line_stats(
+        &reader_bytes,
+        MIN_ROWS_PER_THREAD,
+        eol_char,
+        None,
+        separator,
+        quote_char,
+    )
+    .map(|(mean, std)| {
+        let n_rows = (reader_bytes.len() as f32 / (mean - 0.01 * std)) as usize;
+        (n_rows / MIN_ROWS_PER_THREAD).clamp(1, max_threads)
+    })
+    .unwrap_or(1);
+
+    let file_chunks = get_file_chunks(
+        &reader_bytes,
+        n_threads,
+        None,
+        separator,
+        quote_char,
+        eol_char,
+    );
+
+    POOL.install(|| {
+        file_chunks
```

Comment:
Yeap, I think this is it.

On nit: we found that we often did accidentally `collect`-ed the iterator outside the `POOL`. So I want to enforce a style pattern that reduces the chance of us doing that.

```rust

let iter = collection.par_iter(..);

let result = POOL.install(|| iter.collect());

// in this case
let result = POOL.install(|| iter.sum());

```

Reply:
I don't know if that's hard in Rust type-wise, but if it isn't, how about adding an API to the pool that makes it hard to make this mistake?

```
POOL.run(collection.par_iter(...).collect)
```

Or at least for the collect calls have

```
POOL.collect(collection.par_iter(...))
```

Reply:
We could do that as well. Though I think that should be a follow up PR.